File: src/macos/LanMessenger/Core/Networking/Media/MediaControlMessage.swift
import Foundation

// The control sub-channel: channel 0, UTF-8 JSON objects, each with a `t`
// discriminator, encrypted like any other media payload.
//
// PROTOCOL.md specifies the table; this is the codec for it. Everything the two
// sides say to each other that is not a picture or a keystroke goes through
// here — key confirmation, the picture size, keyframe requests, the control
// escalation, display selection, host state, and the keepalive round trip.
//
// Three rules shape the implementation, and all three are about a peer being
// older, newer, or hostile rather than about the happy path:
//
//  * **An unknown `t` decodes rather than throws.** The whole point of a
//    discriminated control channel is that it can grow. A receiver that faults
//    on a message it has not heard of turns every future extension into a
//    flag-day upgrade, and `PacketValidator` already taught this lesson once —
//    silently dropping unknown types is what made the `caps` field necessary.
//  * **Payloads are bounded.** A control frame is a few hundred bytes. The media
//    frame cap is 4 MiB, so without a second limit here a peer can make a host
//    parse four megabytes of JSON per frame, on the session's read queue, for
//    free.
//  * **Numbers are clamped where they are used, not trusted where they arrive.**
//    A `video_config` claiming 2,000,000 pixels wide is not a resolution, it is
//    an allocation request.

/// Picture geometry. Must arrive before the first video frame and again after
/// any resolution or display change — a viewer has no other source for it, and
/// it is what input coordinates are normalized against.
struct VideoConfig: Codable, Equatable {
    var width: Int
    var height: Int
    /// Backing scale of the captured display, for a viewer that wants to present
    /// at native density. Informational: coordinates are normalized regardless.
    var scale: Double
    var displayID: UInt32
    var codec: String

    init(width: Int, height: Int, scale: Double = 1, displayID: UInt32 = 0,
         codec: String = "h264") {
        self.width = width
        self.height = height
        self.scale = scale
        self.displayID = displayID
        self.codec = codec
    }

    enum CodingKeys: String, CodingKey {
        case width, height, scale, codec
        case displayID = "display_id"
    }

    /// The largest picture either platform will agree to describe.
    ///
    /// 8K-ish, well past any real display, and small enough that a peer cannot
    /// use the field as an allocation primitive. A config outside this is not a
    /// display that exists.
    static let maxDimension = 16_384

    var isPlausible: Bool {
        width > 0 && height > 0
            && width <= Self.maxDimension && height <= Self.maxDimension
            && scale > 0 && scale <= 8
    }
}

/// One of the host's displays, for the picker.
struct RemoteDisplayInfo: Codable, Equatable {
    var displayID: UInt32
    var width: Int
    var height: Int
    var isPrimary: Bool
    var name: String

    init(displayID: UInt32, width: Int, height: Int, isPrimary: Bool, name: String) {
        self.displayID = displayID
        self.width = width
        self.height = height
        self.isPrimary = isPrimary
        self.name = name
    }

    enum CodingKeys: String, CodingKey {
        case width, height, name
        case displayID = "display_id"
        case isPrimary = "is_primary"
    }
}

/// What turns an inexplicable frozen image into an explanation.
///
/// A Windows host cannot capture or drive the secure desktop, and cannot inject
/// into a focused elevated window. Neither is a bug and neither is fixable from
/// user mode, so the viewer is told and shows a banner rather than a mystery.
struct RemoteHostState: Codable, Equatable {
    var secureDesktop: Bool
    var elevatedFocus: Bool
    var locked: Bool

    init(secureDesktop: Bool = false, elevatedFocus: Bool = false, locked: Bool = false) {
        self.secureDesktop = secureDesktop
        self.elevatedFocus = elevatedFocus
        self.locked = locked
    }

    enum CodingKeys: String, CodingKey {
        case secureDesktop = "secure_desktop"
        case elevatedFocus = "elevated_focus"
        case locked
    }

    /// Whether anything is worth saying. All-false is the ordinary state and
    /// must not produce a banner.
    var isNotable: Bool { secureDesktop || elevatedFocus || locked }
}

/// What the stats sub-channel carries. Sent by both sides; each reports what it
/// can see, and neither infers the other's numbers.
struct RemoteSessionStats: Codable, Equatable {
    var roundTripMs: Int
    var decodedFps: Int
    var droppedFrames: Int
    var decodeQueueDepth: Int
    /// Glass-to-glass, derived from `capture_us`. Negative is possible when the
    /// two machines' clocks disagree, and is reported rather than hidden — a
    /// nonsense latency figure is a clock problem worth seeing.
    var endToEndLatencyMs: Int

    init(roundTripMs: Int = 0, decodedFps: Int = 0, droppedFrames: Int = 0,
         decodeQueueDepth: Int = 0, endToEndLatencyMs: Int = 0) {
        self.roundTripMs = roundTripMs
        self.decodedFps = decodedFps
        self.droppedFrames = droppedFrames
        self.decodeQueueDepth = decodeQueueDepth
        self.endToEndLatencyMs = endToEndLatencyMs
    }

    enum CodingKeys: String, CodingKey {
        case roundTripMs = "rtt_ms"
        case decodedFps = "decoded_fps"
        case droppedFrames = "dropped_frames"
        case decodeQueueDepth = "decode_queue"
        case endToEndLatencyMs = "latency_ms"
    }
}

enum MediaControlMessage: Equatable {
    case hello(transcriptB64: String)
    case helloAck
    case videoConfig(VideoConfig)
    case keyframeRequest(reason: String)
    case controlRequest
    case controlGrant
    case controlRevoke
    case displayList([RemoteDisplayInfo])
    case displaySelect(displayID: UInt32)
    case hostState(RemoteHostState)
    case ping(id: UInt64, sentUs: UInt64)
    case pong(id: UInt64, sentUs: UInt64)
    case stats(RemoteSessionStats)
    /// A message this build does not know. Carried rather than dropped so a log
    /// can say what arrived, and so the read loop keeps going.
    case unknown(type: String)

    /// The wire value of `t`.
    var type: String {
        switch self {
        case .hello:           return "hello"
        case .helloAck:        return "hello_ack"
        case .videoConfig:     return "video_config"
        case .keyframeRequest: return "keyframe_request"
        case .controlRequest:  return "control_request"
        case .controlGrant:    return "control_grant"
        case .controlRevoke:   return "control_revoke"
        case .displayList:     return "display_list"
        case .displaySelect:   return "display_select"
        case .hostState:       return "host_state"
        case .ping:            return "ping"
        case .pong:            return "pong"
        case .stats:           return "stats"
        case .unknown(let t):  return t
        }
    }
}

enum MediaControlCodecError: Error, Equatable, CustomStringConvertible {
    case tooLarge(Int)
    case notJSON
    case missingDiscriminator
    case malformed(String)

    var description: String {
        switch self {
        case .tooLarge(let bytes): return "control payload of \(bytes) bytes exceeds the cap"
        case .notJSON:             return "control payload is not a JSON object"
        case .missingDiscriminator: return "control payload has no `t`"
        case .malformed(let type): return "control payload for `\(type)` is malformed"
        }
    }
}

enum MediaControlCodec {

    /// A control frame is a few hundred bytes. The media frame cap is 4 MiB, so
    /// without a second limit a peer can make a host parse four megabytes of
    /// JSON per frame on the session's read queue, for free.
    ///
    /// 64 KiB leaves ample room for a `display_list` on a machine with an
    /// improbable number of monitors, and nothing else here is near it.
    static let maxPayloadBytes = 64 * 1024

    // MARK: - Encode

    static func encode(_ message: MediaControlMessage) throws -> Data {
        var object: [String: Any] = ["t": message.type]

        switch message {
        case .hello(let transcript):
            object["transcript_b64"] = transcript
        case .videoConfig(let config):
            object["config"] = try dictionary(from: config)
        case .keyframeRequest(let reason):
            object["reason"] = reason
        case .displayList(let displays):
            object["displays"] = try displays.map { try dictionary(from: $0) }
        case .displaySelect(let displayID):
            object["display_id"] = displayID
        case .hostState(let state):
            object["state"] = try dictionary(from: state)
        case .ping(let id, let sentUs), .pong(let id, let sentUs):
            object["id"] = id
            object["sent_us"] = sentUs
        case .stats(let stats):
            object["stats"] = try dictionary(from: stats)
        case .helloAck, .controlRequest, .controlGrant, .controlRevoke, .unknown:
            break
        }

        // Sorted keys so the two platforms produce byte-identical output for the
        // same message, which is what makes a shared conformance vector possible.
        let data = try JSONSerialization.data(withJSONObject: object,
                                              options: [.sortedKeys])
        guard data.count <= maxPayloadBytes else {
            throw MediaControlCodecError.tooLarge(data.count)
        }
        return data
    }

    // MARK: - Decode

    static func decode(_ data: Data) throws -> MediaControlMessage {
        guard data.count <= maxPayloadBytes else {
            throw MediaControlCodecError.tooLarge(data.count)
        }
        guard let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw MediaControlCodecError.notJSON
        }
        guard let type = object["t"] as? String, !type.isEmpty else {
            throw MediaControlCodecError.missingDiscriminator
        }

        switch type {
        case "hello":
            return .hello(transcriptB64: object["transcript_b64"] as? String ?? "")
        case "hello_ack":
            return .helloAck
        case "video_config":
            guard let raw = object["config"],
                  let config: VideoConfig = try? value(from: raw) else {
                throw MediaControlCodecError.malformed(type)
            }
            return .videoConfig(config)
        case "keyframe_request":
            return .keyframeRequest(reason: object["reason"] as? String ?? "")
        case "control_request":
            return .controlRequest
        case "control_grant":
            return .controlGrant
        case "control_revoke":
            return .controlRevoke
        case "display_list":
            guard let raw = object["displays"] as? [Any] else {
                throw MediaControlCodecError.malformed(type)
            }
            // A single unparseable entry drops that display rather than the
            // whole list: losing one monitor from a picker is recoverable,
            // losing the picker is not.
            let displays = raw.compactMap { try? value(from: $0) as RemoteDisplayInfo }
            return .displayList(displays)
        case "display_select":
            guard let id = object["display_id"] as? NSNumber else {
                throw MediaControlCodecError.malformed(type)
            }
            return .displaySelect(displayID: id.uint32Value)
        case "host_state":
            guard let raw = object["state"],
                  let state: RemoteHostState = try? value(from: raw) else {
                throw MediaControlCodecError.malformed(type)
            }
            return .hostState(state)
        case "ping", "pong":
            let id = (object["id"] as? NSNumber)?.uint64Value ?? 0
            let sent = (object["sent_us"] as? NSNumber)?.uint64Value ?? 0
            return type == "ping" ? .ping(id: id, sentUs: sent) : .pong(id: id, sentUs: sent)
        case "stats":
            guard let raw = object["stats"],
                  let stats: RemoteSessionStats = try? value(from: raw) else {
                throw MediaControlCodecError.malformed(type)
            }
            return .stats(stats)
        default:
            // The whole point of a discriminated channel is that it can grow.
            // Faulting here would turn every future extension into a flag-day
            // upgrade — the lesson `PacketValidator` already taught once.
            return .unknown(type: type)
        }
    }

    // MARK: - Private

    private static func dictionary<T: Encodable>(from value: T) throws -> [String: Any] {
        let data = try JSONEncoder().encode(value)
        guard let object = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw MediaControlCodecError.malformed("\(T.self)")
        }
        return object
    }

    /// The type check is not a formality.
    ///
    /// `JSONSerialization.data(withJSONObject:)` raises an **ObjC exception**
    /// rather than throwing a Swift error when handed anything that is not a
    /// top-level container — a bare string, a number, null. `try?` does not
    /// catch that, so a peer sending `"displays": ["nonsense"]` would take the
    /// session's read loop down with it. Tolerant-looking code that calls this
    /// without checking first is a remote crash, not a lenient parser.
    private static func value<T: Decodable>(from raw: Any) throws -> T {
        guard JSONSerialization.isValidJSONObject([raw]) , raw is [String: Any] else {
            throw MediaControlCodecError.malformed("\(T.self)")
        }
        let data = try JSONSerialization.data(withJSONObject: raw)
        return try JSONDecoder().decode(T.self, from: data)
    }
}

File: src/windows-native/LanMessenger/Core/Networking/Media/MediaControlMessage.cs
using System.Text.Encodings.Web;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace LanMessenger.Core.Networking.Media;

// The control sub-channel: channel 0, UTF-8 JSON objects, each with a `t`
// discriminator. Mirror of the macOS MediaControlMessage.swift.
//
// The two encoders must produce byte-identical output for the same message,
// which is what makes a shared conformance vector possible. Two things get in
// the way of that on this side and both are handled here:
//
//   * System.Text.Json writes properties in declaration order and has no
//     sort option, so every object is written by hand with Utf8JsonWriter in
//     sorted key order to match Foundation's .sortedKeys.
//   * System.Text.Json escapes '/' and every non-ASCII character by default,
//     where Foundation does not. UnsafeRelaxedJsonEscaping turns that off. The
//     same trap bit the handshake transcript, which is why that is hand-rolled
//     too — one byte of difference and the two sides disagree.

public sealed class VideoConfig
{
    [JsonPropertyName("width")]      public int    Width     { get; set; }
    [JsonPropertyName("height")]     public int    Height    { get; set; }
    [JsonPropertyName("scale")]      public double Scale     { get; set; } = 1;
    [JsonPropertyName("display_id")] public uint   DisplayId { get; set; }
    [JsonPropertyName("codec")]      public string Codec     { get; set; } = "h264";

    /// The largest picture either platform will agree to describe. Well past any
    /// real display, and small enough that a peer cannot use the field as an
    /// allocation primitive.
    public const int MaxDimension = 16384;

    [JsonIgnore]
    public bool IsPlausible =>
        Width > 0 && Height > 0 && Width <= MaxDimension && Height <= MaxDimension
        && Scale > 0 && Scale <= 8;

    public override bool Equals(object? obj) =>
        obj is VideoConfig o && o.Width == Width && o.Height == Height
        && o.Scale.Equals(Scale) && o.DisplayId == DisplayId && o.Codec == Codec;

    public override int GetHashCode() => HashCode.Combine(Width, Height, Scale, DisplayId, Codec);
}

public sealed class RemoteDisplayInfo
{
    [JsonPropertyName("display_id")] public uint   DisplayId { get; set; }
    [JsonPropertyName("width")]      public int    Width     { get; set; }
    [JsonPropertyName("height")]     public int    Height    { get; set; }
    [JsonPropertyName("is_primary")] public bool   IsPrimary { get; set; }
    [JsonPropertyName("name")]       public string Name      { get; set; } = "";

    public override bool Equals(object? obj) =>
        obj is RemoteDisplayInfo o && o.DisplayId == DisplayId && o.Width == Width
        && o.Height == Height && o.IsPrimary == IsPrimary && o.Name == Name;

    public override int GetHashCode() => HashCode.Combine(DisplayId, Width, Height, IsPrimary, Name);
}

/// What turns an inexplicable frozen image into an explanation. A Windows host
/// cannot capture or drive the secure desktop, and cannot inject into a focused
/// elevated window; neither is a bug and neither is fixable from user mode, so
/// the viewer is told and shows a banner rather than a mystery.
public sealed class RemoteHostState
{
    [JsonPropertyName("secure_desktop")] public bool SecureDesktop { get; set; }
    [JsonPropertyName("elevated_focus")] public bool ElevatedFocus { get; set; }
    [JsonPropertyName("locked")]         public bool Locked        { get; set; }

    [JsonIgnore]
    public bool IsNotable => SecureDesktop || ElevatedFocus || Locked;

    public override bool Equals(object? obj) =>
        obj is RemoteHostState o && o.SecureDesktop == SecureDesktop
        && o.ElevatedFocus == ElevatedFocus && o.Locked == Locked;

    public override int GetHashCode() => HashCode.Combine(SecureDesktop, ElevatedFocus, Locked);
}

public sealed class RemoteSessionStats
{
    [JsonPropertyName("rtt_ms")]        public int RoundTripMs       { get; set; }
    [JsonPropertyName("decoded_fps")]   public int DecodedFps        { get; set; }
    [JsonPropertyName("dropped_frames")] public int DroppedFrames    { get; set; }
    [JsonPropertyName("decode_queue")]  public int DecodeQueueDepth  { get; set; }
    /// Glass-to-glass, derived from capture_us. Negative is possible when the two
    /// machines' clocks disagree, and is reported rather than hidden — a nonsense
    /// latency figure is a clock problem worth seeing.
    [JsonPropertyName("latency_ms")]    public int EndToEndLatencyMs { get; set; }

    public override bool Equals(object? obj) =>
        obj is RemoteSessionStats o && o.RoundTripMs == RoundTripMs
        && o.DecodedFps == DecodedFps && o.DroppedFrames == DroppedFrames
        && o.DecodeQueueDepth == DecodeQueueDepth && o.EndToEndLatencyMs == EndToEndLatencyMs;

    public override int GetHashCode() =>
        HashCode.Combine(RoundTripMs, DecodedFps, DroppedFrames, DecodeQueueDepth, EndToEndLatencyMs);
}

public enum MediaControlType
{
    Hello, HelloAck, VideoConfig, KeyframeRequest, ControlRequest, ControlGrant,
    ControlRevoke, DisplayList, DisplaySelect, HostState, Ping, Pong, Stats, Unknown,
}

public sealed class MediaControlMessage
{
    public MediaControlType Kind { get; init; }
    /// The wire value of `t`. Carried verbatim so an unknown message can be
    /// logged by name.
    public string Type { get; init; } = "";

    public string TranscriptB64 { get; init; } = "";
    public VideoConfig? Config { get; init; }
    public string Reason { get; init; } = "";
    public List<RemoteDisplayInfo> Displays { get; init; } = [];
    public uint DisplayId { get; init; }
    public RemoteHostState? State { get; init; }
    public ulong Id { get; init; }
    public ulong SentUs { get; init; }
    public RemoteSessionStats? Stats { get; init; }

    public static MediaControlMessage Hello(string transcriptB64) =>
        new() { Kind = MediaControlType.Hello, Type = "hello", TranscriptB64 = transcriptB64 };
    public static MediaControlMessage HelloAck() =>
        new() { Kind = MediaControlType.HelloAck, Type = "hello_ack" };
    public static MediaControlMessage Video(VideoConfig config) =>
        new() { Kind = MediaControlType.VideoConfig, Type = "video_config", Config = config };
    public static MediaControlMessage KeyframeRequest(string reason) =>
        new() { Kind = MediaControlType.KeyframeRequest, Type = "keyframe_request", Reason = reason };
    public static MediaControlMessage ControlRequest() =>
        new() { Kind = MediaControlType.ControlRequest, Type = "control_request" };
    public static MediaControlMessage ControlGrant() =>
        new() { Kind = MediaControlType.ControlGrant, Type = "control_grant" };
    public static MediaControlMessage ControlRevoke() =>
        new() { Kind = MediaControlType.ControlRevoke, Type = "control_revoke" };
    public static MediaControlMessage DisplayList(List<RemoteDisplayInfo> displays) =>
        new() { Kind = MediaControlType.DisplayList, Type = "display_list", Displays = displays };
    public static MediaControlMessage DisplaySelect(uint displayId) =>
        new() { Kind = MediaControlType.DisplaySelect, Type = "display_select", DisplayId = displayId };
    public static MediaControlMessage Host(RemoteHostState state) =>
        new() { Kind = MediaControlType.HostState, Type = "host_state", State = state };
    public static MediaControlMessage Ping(ulong id, ulong sentUs) =>
        new() { Kind = MediaControlType.Ping, Type = "ping", Id = id, SentUs = sentUs };
    public static MediaControlMessage Pong(ulong id, ulong sentUs) =>
        new() { Kind = MediaControlType.Pong, Type = "pong", Id = id, SentUs = sentUs };
    public static MediaControlMessage Statistics(RemoteSessionStats stats) =>
        new() { Kind = MediaControlType.Stats, Type = "stats", Stats = stats };
    public static MediaControlMessage Unknown(string type) =>
        new() { Kind = MediaControlType.Unknown, Type = type };
}

public sealed class MediaControlCodecException(string message) : Exception(message);

public static class MediaControlCodec
{
    /// A control frame is a few hundred bytes. The media frame cap is 4 MiB, so
    /// without a second limit a peer can make a host parse four megabytes of JSON
    /// per frame on the session's read queue, for free.
    public const int MaxPayloadBytes = 64 * 1024;

    private static readonly JsonWriterOptions WriterOptions = new()
    {
        // Foundation does not escape '/' or non-ASCII; System.Text.Json does
        // both by default. One byte of difference and the two sides disagree.
        Encoder = JavaScriptEncoder.UnsafeRelaxedJsonEscaping,
        Indented = false,
    };

    // ---- Encode ------------------------------------------------------------

    public static byte[] Encode(MediaControlMessage message)
    {
        using var stream = new MemoryStream();
        using (var w = new Utf8JsonWriter(stream, WriterOptions))
        {
            w.WriteStartObject();
            // Sorted key order throughout, by hand, to match Foundation's
            // .sortedKeys. System.Text.Json writes declaration order and offers
            // no sort option.
            switch (message.Kind)
            {
                case MediaControlType.VideoConfig:
                    WriteConfig(w, "config", message.Config ?? new VideoConfig());
                    w.WriteString("t", message.Type);
                    break;
                case MediaControlType.DisplaySelect:
                    w.WriteNumber("display_id", message.DisplayId);
                    w.WriteString("t", message.Type);
                    break;
                case MediaControlType.DisplayList:
                    w.WriteStartArray("displays");
                    foreach (var display in message.Displays) WriteDisplay(w, display);
                    w.WriteEndArray();
                    w.WriteString("t", message.Type);
                    break;
                case MediaControlType.Ping:
                case MediaControlType.Pong:
                    w.WriteNumber("id", message.Id);
                    w.WriteNumber("sent_us", message.SentUs);
                    w.WriteString("t", message.Type);
                    break;
                case MediaControlType.KeyframeRequest:
                    w.WriteString("reason", message.Reason);
                    w.WriteString("t", message.Type);
                    break;
                case MediaControlType.HostState:
                    WriteState(w, "state", message.State ?? new RemoteHostState());
                    w.WriteString("t", message.Type);
                    break;
                case MediaControlType.Stats:
                    WriteStats(w, "stats", message.Stats ?? new RemoteSessionStats());
                    w.WriteString("t", message.Type);
                    break;
                case MediaControlType.Hello:
                    w.WriteString("t", message.Type);
                    w.WriteString("transcript_b64", message.TranscriptB64);
                    break;
                default:
                    w.WriteString("t", message.Type);
                    break;
            }
            w.WriteEndObject();
        }

        var bytes = stream.ToArray();
        if (bytes.Length > MaxPayloadBytes)
        {
            throw new MediaControlCodecException(
                $"control payload of {bytes.Length} bytes exceeds the cap");
        }
        return bytes;
    }

    private static void WriteConfig(Utf8JsonWriter w, string name, VideoConfig c)
    {
        w.WriteStartObject(name);
        w.WriteString("codec", c.Codec);
        w.WriteNumber("display_id", c.DisplayId);
        w.WriteNumber("height", c.Height);
        WriteScale(w, "scale", c.Scale);
        w.WriteNumber("width", c.Width);
        w.WriteEndObject();
    }

    private static void WriteDisplay(Utf8JsonWriter w, RemoteDisplayInfo d)
    {
        w.WriteStartObject();
        w.WriteNumber("display_id", d.DisplayId);
        w.WriteNumber("height", d.Height);
        w.WriteBoolean("is_primary", d.IsPrimary);
        w.WriteString("name", d.Name);
        w.WriteNumber("width", d.Width);
        w.WriteEndObject();
    }

    private static void WriteState(Utf8JsonWriter w, string name, RemoteHostState s)
    {
        w.WriteStartObject(name);
        w.WriteBoolean("elevated_focus", s.ElevatedFocus);
        w.WriteBoolean("locked", s.Locked);
        w.WriteBoolean("secure_desktop", s.SecureDesktop);
        w.WriteEndObject();
    }

    private static void WriteStats(Utf8JsonWriter w, string name, RemoteSessionStats s)
    {
        w.WriteStartObject(name);
        w.WriteNumber("decode_queue", s.DecodeQueueDepth);
        w.WriteNumber("decoded_fps", s.DecodedFps);
        w.WriteNumber("dropped_frames", s.DroppedFrames);
        w.WriteNumber("latency_ms", s.EndToEndLatencyMs);
        w.WriteNumber("rtt_ms", s.RoundTripMs);
        w.WriteEndObject();
    }

    /// Foundation writes a whole-valued Double without a decimal point, so 1.0
    /// becomes `1`. Matching that is what keeps the encoded bytes identical.
    private static void WriteScale(Utf8JsonWriter w, string name, double value)
    {
        if (value == Math.Floor(value) && Math.Abs(value) < 1e15)
        {
            w.WriteNumber(name, (long)value);
        }
        else
        {
            w.WriteNumber(name, value);
        }
    }

    // ---- Decode ------------------------------------------------------------

    public static MediaControlMessage Decode(ReadOnlySpan<byte> data)
    {
        if (data.Length > MaxPayloadBytes)
        {
            throw new MediaControlCodecException(
                $"control payload of {data.Length} bytes exceeds the cap");
        }

        JsonDocument document;
        try { document = JsonDocument.Parse(data.ToArray()); }
        catch (JsonException) { throw new MediaControlCodecException("control payload is not JSON"); }

        using (document)
        {
            var root = document.RootElement;
            if (root.ValueKind != JsonValueKind.Object)
            {
                throw new MediaControlCodecException("control payload is not a JSON object");
            }
            if (!root.TryGetProperty("t", out var typeElement) ||
                typeElement.ValueKind != JsonValueKind.String ||
                string.IsNullOrEmpty(typeElement.GetString()))
            {
                throw new MediaControlCodecException("control payload has no `t`");
            }

            var type = typeElement.GetString()!;
            switch (type)
            {
                case "hello":
                    return MediaControlMessage.Hello(GetString(root, "transcript_b64"));
                case "hello_ack":
                    return MediaControlMessage.HelloAck();
                case "video_config":
                    return MediaControlMessage.Video(
                        Deserialize<VideoConfig>(root, "config", type));
                case "keyframe_request":
                    return MediaControlMessage.KeyframeRequest(GetString(root, "reason"));
                case "control_request":
                    return MediaControlMessage.ControlRequest();
                case "control_grant":
                    return MediaControlMessage.ControlGrant();
                case "control_revoke":
                    return MediaControlMessage.ControlRevoke();
                case "display_list":
                {
                    if (!root.TryGetProperty("displays", out var list) ||
                        list.ValueKind != JsonValueKind.Array)
                    {
                        throw new MediaControlCodecException("`display_list` is malformed");
                    }
                    // A single unparseable entry drops that display rather than
                    // the whole list: losing one monitor from a picker is
                    // recoverable, losing the picker is not.
                    var displays = new List<RemoteDisplayInfo>();
                    foreach (var item in list.EnumerateArray())
                    {
                        try
                        {
                            var parsed = item.Deserialize<RemoteDisplayInfo>();
                            if (parsed is not null) displays.Add(parsed);
                        }
                        catch (JsonException) { }
                    }
                    return MediaControlMessage.DisplayList(displays);
                }
                case "display_select":
                {
                    if (!root.TryGetProperty("display_id", out var id) ||
                        !id.TryGetUInt32(out var value))
                    {
                        throw new MediaControlCodecException("`display_select` is malformed");
                    }
                    return MediaControlMessage.DisplaySelect(value);
                }
                case "host_state":
                    return MediaControlMessage.Host(
                        Deserialize<RemoteHostState>(root, "state", type));
                case "ping":
                case "pong":
                {
                    var id = root.TryGetProperty("id", out var i) && i.TryGetUInt64(out var iv) ? iv : 0;
                    var sent = root.TryGetProperty("sent_us", out var s) && s.TryGetUInt64(out var sv) ? sv : 0;
                    return type == "ping"
                        ? MediaControlMessage.Ping(id, sent)
                        : MediaControlMessage.Pong(id, sent);
                }
                case "stats":
                    return MediaControlMessage.Statistics(
                        Deserialize<RemoteSessionStats>(root, "stats", type));
                default:
                    // The whole point of a discriminated channel is that it can
                    // grow. Faulting here would turn every future extension into
                    // a flag-day upgrade.
                    return MediaControlMessage.Unknown(type);
            }
        }
    }

    private static string GetString(JsonElement root, string name) =>
        root.TryGetProperty(name, out var e) && e.ValueKind == JsonValueKind.String
            ? e.GetString() ?? "" : "";

    private static T Deserialize<T>(JsonElement root, string name, string type) where T : class
    {
        if (!root.TryGetProperty(name, out var element))
        {
            throw new MediaControlCodecException($"`{type}` is malformed");
        }
        try
        {
            return element.Deserialize<T>()
                   ?? throw new MediaControlCodecException($"`{type}` is malformed");
        }
        catch (JsonException)
        {
            throw new MediaControlCodecException($"`{type}` is malformed");
        }
    }
}
